### ISSUE TITLE
Internal FileCache should allow configuration of max age and/or max disk space

### ISSUE THREAD

**Issue by akitchen** [mapbox/mapbox-navigation-ios#1384]
As we add support for caching more types of resources in the SDK we should start being more intentional about the file cache footprint.

We should introduce support at least for maximum disk space configuration; when we are near the maximum we'll need to figure out how to be smart about what objects to evict (the oldest objects aren't necessarily the least used).

A notion of maximum age for a cached resource could also be considered, configurable on a pre-item basis whenever an item is put into the cache.

If we implement both, then max space eviction could prefer objects closest to their expiration date.

**Comment by vincethecoder** [comment]
In lieu of satisfying the _maximum disk space configuration, maximum age for a cached resource, whenever an item is put into the cache_, we could adopt the following strategies...

> 
>  [maxDiskCacheSize](#)
> _The largest disk size can be taken for the cache. It is the total allocated size of the file in bytes. Default is 0, which means no limit._
> ```swift 
> public var maxDiskCacheSize: UInt = 0
> ```
> 
> [maxCachePeriodInSeconds](#)
> _The longest time duration of the cache being stored in disk. Default is 1 week._
> ```swift
> public var maxMemoryCost: UInt = 0
> ```


_When we are near the maximum, we could potentially rely on an internal `refCount` and delete items which have been barely accessed_
> `image.refCount`: default 0 upon cache and `cachedImageForKey`: increment refCount by.

**Developer's configuration**:
```swift
cache.maxCachePeriodInSeconds = YOUR_VALUE
cache.maxDiskCacheSize = YOUR_VALUE
```
Reference: [Kingfisher ImageCache Class Reference ](http://cocoadocs.org/docsets/Kingfisher/1.1.2/Classes/ImageCache.html#/s:vC10Kingfisher10ImageCache13maxMemoryCostSu)

###### Future efforts, we could also consider the largest cache cost of memory cache per item stored.

**Comment by bsudekum** [comment]
This sounds good to me. One suggestion, to guard against a crazy situation where either a bug/long route/edge case produces gigabytes of cached images or sound files, it might make sense to have a rather large default instead of unlimited. I think it would only be necessary to configure `maxDiskCacheSize`.  Maybe something like 2gb?

**Comment by JThramer** [comment]
Posterity: It seems to me that the proper way to implement this would to be to keep a "metadata array" around in the cache that describes the cached elements. Maybe an array of type T, where T contains filesize, timestamp, and a reference to the cached item (weak pointer for memory cache, file URL for disk cache), which will allow us to check for oversize/outdated elements. The tricky part about this is that in the `DiskCache`, the metadata array will _also_ need to be serialized so it persists between runs.

**Comment by akitchen** [comment]
This might make sense, and @1ec5 offered a similar suggestion some weeks back (citation needed)

I don't think we'd need to keep a pointer/reference, but we do need a unique identifier to key off of (the `cacheKey` perhaps?). 

Also, if we're going down this path, I'd like to add a `last_accessed_at` time stamp to enable prioritizing least recently accessed objects for eviction.

All of this feels like it might be over-engineering if the footprint is small; we might want to take a closer look at the cache usage across long-running app installs. Simply deleting old objects based on the creation-time filesystem timestamp might be enough... there is a 'FileAttributeKey` for that, and of course size as well, so we might be able to get away with not introducing a metadata table unless we want to implement LRU eviction.

**Comment by 1ec5** [comment]
>> It seems to me that the proper way to implement this would to be to keep a "metadata array" around in the cache that describes the cached elements.
> 
> This might make sense, and @1ec5 offered a similar suggestion some weeks back (citation needed)

Yes, this is what I had in mind: https://github.com/mapbox/mapbox-navigation-ios/issues/1391#issuecomment-387507267.
